src/Appwrite/Certificates/Adapter/LetsEncrypt.php

```diff
@@ -0,0 +1,126 @@
+            throw new Exception('Failed to rename certificate chain.pem. Let\'s Encrypt log: ' . $stderr . ' ; ' . $stdout);
+        }
+
+        if (!@\rename('/etc/letsencrypt/live/' . $certName . '/fullchain.pem', APP_STORAGE_CERTIFICATES . '/' . $domain . '/fullchain.pem')) {
+            throw new Exception('Failed to rename certificate fullchain.pem. Let\'s Encrypt log: ' . $stderr . ' ; ' . $stdout);
+        }
+
+        if (!@\rename('/etc/letsencrypt/live/' . $certName . '/privkey.pem', APP_STORAGE_CERTIFICATES . '/' . $domain . '/privkey.pem')) {
+            throw new Exception('Failed to rename certificate privkey.pem. Let\'s Encrypt log: ' . $stderr . ' ; ' . $stdout);
+        }
+
+        $config = \implode(PHP_EOL, [
+            "tls:",
+            "  certificates:",
+            "    - certFile: /storage/certificates/{$domain}/fullchain.pem",
+            "      keyFile: /storage/certificates/{$domain}/privkey.pem"
+        ]);
+
+        // Save configuration into Traefik using our new cert files
+        if (!\file_put_contents(APP_STORAGE_CONFIG . '/' . $domain . '.yml', $config)) {
+            throw new Exception('Failed to save Traefik configuration.');
+        }
+
+        $certPath = APP_STORAGE_CERTIFICATES . '/' . $domain . '/cert.pem';
+        $certData = openssl_x509_parse(file_get_contents($certPath));
+        $validTo = $certData['validTo_time_t'] ?? null;
+        $dt = (new \DateTime())->setTimestamp($validTo);
+        return DateTime::addSeconds($dt, -60 * 60 * 24 * 30);
+    }
+
+    public function isRenewRequired(string $domain, Log $log): bool
+    {
+        $certPath = APP_STORAGE_CERTIFICATES . '/' . $domain . '/cert.pem';
+        if (\file_exists($certPath)) {
+            $certData = openssl_x509_parse(file_get_contents($certPath));
+            $validTo = $certData['validTo_time_t'] ?? 0;
+
+            if (empty($validTo)) {
+                $log->addTag('certificateDomain', $domain);
+                throw new Exception('Unable to read certificate file (cert.pem).');
+            }
+
+            // LetsEncrypt allows renewal 30 days before expiry
+            $expiryInAdvance = (60 * 60 * 24 * 30);
+            if ($validTo - $expiryInAdvance > \time()) {
+                $log->addTag('certificateDomain', $domain);
+                $log->addExtra('certificateData', \is_array($certData) ? \json_encode($certData) : \strval($certData));
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public function deleteCertificate(string $domain): void
+    {
+        $directory = APP_STORAGE_CERTIFICATES . '/' . $domain;
+        $checkTraversal = realpath($directory) === $directory;
+
+        if ($checkTraversal && is_dir($directory)) {
+            // Delete files, so Traefik is aware of change
+            array_map('unlink', glob($directory . '/*.*'));
+            rmdir($directory);
```

Comment:
This is a good chance to start using Utopia/Storage for all file interactions to better follow our conventions. Missing functions can be added.